File: src/main/java/com/vc/medium/L159.scala
package com.vc.hard

object L159 {
  def lengthOfLongestSubstringTwoDistinct(s: String): Int = {
    import scala.collection.mutable
    val map = new mutable.HashMap[Char, Int]()
    var start = 0
    var end = 0
    var res = 0
    while(end < s.length) {
      val add = s(end)
      map.put(add, map.getOrElse(add, 0) + 1)

      while(map.size > 2) {
        val remove = s(start)
        if(map(remove) - 1 == 0) map.remove(remove)
        else map.put(remove, map(remove) - 1)
        start += 1
      }

      res = Math.max(res, end - start + 1)
      end += 1
    }
    res
  }
}

File: src/main/java/com/vc/medium/L159.java
package com.vc.medium;

import java.util.HashMap;

class L159 {
    public int lengthOfLongestSubstringTwoDistinct(String s) {
        HashMap<Character, Integer> map = new HashMap<>();
        int start = 0, end = 0, res = 0, n = s.length(), k = 2;
        while(end < n) {
            char ch = s.charAt(end);
            map.put(ch, map.getOrDefault(ch, 0) + 1);
            while(start < s.length() && map.size() > k) {
                char toBeRemoved = s.charAt(start++);
                map.put(toBeRemoved, map.getOrDefault(toBeRemoved, 0) - 1);
                if(map.get(toBeRemoved) <= 0) map.remove(toBeRemoved);
            }
            if(map.size() <= k) res = Math.max(res, end - start + 1);
            end++;
        }
        return res;
    }
}
